src/Functions/multiIf.cpp

```diff
@@ -257,11 +320,68 @@ class FunctionMultiIf final : public FunctionIfBase
                 }
             }
         }
+    }
 
-        return res;
+    /// We should read source from which instruction on each row?
+    template <typename S>
+    static void calculateInserts(std::vector<Instruction> & instructions, size_t rows, PaddedPODArray<S> & inserts)
+    {
+        for (S i = static_cast<S>(instructions.size() - 1); i >= 0; --i)
+        {
+            auto & instruction = instructions[i];
+            if (instruction.condition_always_true)
+            {
+                for (size_t row_i = 0; row_i < rows; ++row_i)
+                    inserts[row_i] = i;
+            }
+            else if (!instruction.condition_is_nullable)
+            {
+                const auto & cond_data = assert_cast<const ColumnUInt8 &>(*instruction.condition).getData();
+                for (size_t row_i = 0; row_i < rows; ++row_i)
+                {
+                    size_t condition_index = instruction.condition_is_short ? instruction.condition_index++ : row_i;
+
+                    /// Equivalent to below code. But it is able to utilize SIMD instructions.
+                    /// if (cond_data[condition_index])
+                    ///     inserts[row_i] = i;
+                    inserts[row_i] += cond_data[condition_index] * (i - inserts[row_i]);
+                }
```

Comment:
It will lead to segfault if any instruction condition is short (when short circuit execution is used). If condition is short, it means that it's size is less than the number of rows in the result. Actually, if it's short, it's size is equal to the number of rows in which all previous conditions were false. In not columnar implementation we track correct index in a short condition by incrementing `instruction.condition_index++` only when all previous conditions were false. Here you just increment it every row, so it will lead to segfault. I guess it's not possible to calculate right row indexes for short condition going backwards by instructions.
When at least one condition is short, we should go forward by instructions and track already inserted rows to use correct condition indexes